### PR TITLE
Addition of "Tabulator" tab-state controller

### PR DESCRIPTION
**Pull request by chases2**
Creates a controller that simply copies grid data to a tab-specific location instead of a test-group specific one.

Includes:
- Refactor of configuration snapshot logic (suggested in #775); first commit
- Extension of GCS fake to support GenerationNotMatch
- Bugfix in summarizer's config snapshot

### CODE REVIEW

File: config/snapshot/config_snapshot.go [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+// AllDashboards returns all dashboards
+func (cs *Config) AllDashboards() []*configpb.Dashboard {
+	if cs == nil {
+		return nil
+	}
+	cs.lock.RLock()
+	defer cs.lock.RUnlock()
+
+	db := make([]*configpb.Dashboard, len(cs.dashboards))
+	var i int
```
Comment (michelle192837) [inline config/snapshot/config_snapshot.go:60]: Does i do anything here?

**Review by michelle192837 — APPROVED** [review]
/hold

File: config/snapshot/config_snapshot.go (hunk 1 above)
Comment (chases2) [reply config/snapshot/config_snapshot.go:60]: Yeah, it's used for the index for the loop on the next few lines. This seemed the clearest way to copy the values in a map into a slice.

File: cmd/tabulator/main.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+	trace    bool
+	jsonLogs bool
+}
+
+// validate ensures sane options
+func (o *options) validate() error {
+	if o.config.String() == "" {
+		return errors.New("empty --config")
+	}
+	if o.concurrency == 0 {
```
[the remark this answers is not in the document]
Comment (chases2) [reply cmd/tabulator/main.go]: Done

File: config/snapshot/config_snapshot.go (hunk 1 above)
Comment (michelle192837) [reply config/snapshot/config_snapshot.go:60]: Oh lol, not sure how I missed that.

**Review by michelle192837 — APPROVED** [review]
/hold cancel

File: config/snapshot/config_snapshot_test.go [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+	"errors"
+	"sync"
+	"testing"
+
+	"cloud.google.com/go/storage"
+	configpb "github.com/GoogleCloudPlatform/testgrid/pb/config"
+	"github.com/GoogleCloudPlatform/testgrid/util/gcs"
+	"github.com/GoogleCloudPlatform/testgrid/util/gcs/fake"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/net/context"
```
Comment (fejta) [inline config/snapshot/config_snapshot_test.go:29]: Can you switch this to "context"?

**Pull request merged**